### PR TITLE
[FLINK-20704][table-planner] Some rel data type does not implement th…

### PR DESCRIPTION
**Pull request by danny0405**
…e digest correctly

## What is the purpose of the change

Some of the rel data types for legacy planner:

- `GenericRelDataType`
- `ArrayRelDataType`
- `MapRelDataType`
- `MultisetRelDataType`

Does not implement the digest correctly, especially for
`GenericRelDataType` , the `RelDataTypeFactory` caches the type instances
based on its digest, a wrong digest impl would mess up the type instance
creation, e.g. without this patch, all the `GenericRelDataType` instances
have same digest `ANY`.

## Brief change log

- Fix the digest of `GenericRelDataType`, `ArrayRelDataType`, `MapRelDataType`, `MultisetRelDataType`

## Verifying this change

No tests needed.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

### CODE REVIEW

**Review by godfreyhe — COMMENTED** [review]
is there any tests to cover the changes?

**Comment by danny0405** [comment]
> is there any tests to cover the changes?

Yes, i have added some tests.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/calcite/FlinkTypeFactoryTest.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.apache.calcite.sql.`type`.SqlTypeName
+import org.junit.Assert.{assertFalse, assertTrue}
+import org.junit.Test
+
+/** Test cases for [[FlinkTypeFactory]]. */
+class FlinkTypeFactoryTest {
+
+  @Test
+  def testCanonizeType(): Unit = {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/calcite/FlinkTypeFactoryTest.scala:13]: I find the test does not cover the changes. I revert the changes and run the method, all tests are success.

Comment (danny0405) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/calcite/FlinkTypeFactoryTest.scala:13]: Thanks for the check, i have updated the test cases.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/calcite/FlinkTypeFactoryTest.scala [hunk 2]
```diff
@@ -91,4 +94,16 @@ class FlinkTypeFactoryTest {
...
   }
+
+  @Test
+  def testCanonizeType(): Unit = {
+    val typeFactory = FlinkTypeFactory.INSTANCE
+    val genericTypeInfo = Types.GENERIC(classOf[TestClass])
+    val genericRelType = typeFactory.createFieldTypeFromLogicalType(new TypeInformationRawType(genericTypeInfo))
+    val genericRelType2 = typeFactory.createFieldTypeFromLogicalType(new TypeInformationRawType(genericTypeInfo))
+
+    assertTrue("The type expect to be canonized", genericRelType == genericRelType2)
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/calcite/FlinkTypeFactoryTest.scala:105]: should use `eq` instead of `==` for object equivalence

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/schema/GenericRelDataType.scala [hunk 3]
```diff
@@ -55,4 +53,10 @@ class GenericRelDataType(
   override def hashCode(): Int = {
     typeInfo.hashCode()
   }
+
+  override def isNullable: Boolean = nullable
+
+  override def generateTypeString(sb: java.lang.StringBuilder, withDetail: Boolean): Unit = {
+    sb.append(s"ANY($typeInfo)")
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/schema/GenericRelDataType.scala:60]: nit: use `toString()` ?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/schema/MapRelDataType.scala [hunk 4]
```diff
@@ -46,4 +46,7 @@ class MapRelDataType(
     Objects.hashCode(keyType, valueType)
   }
 
+  override def generateTypeString(sb: java.lang.StringBuilder, withDetail: Boolean): Unit = {
+    sb.append(s"MAP($typeInfo)")
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/schema/MapRelDataType.scala:50]: ditto

**Review by godfreyhe — COMMENTED** [review]
Thanks for the update, LGTM over all, I left a few minor comments

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/schema/MapRelDataType.scala [hunk 5]
```diff
@@ -45,5 +45,4 @@ class MapRelDataType(
   override def hashCode(): Int = {
     Objects.hashCode(keyType, valueType)
   }
-
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/schema/MapRelDataType.scala:48]: the change is reverted ?

Comment (danny0405) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/schema/MapRelDataType.scala:48]: Yes, the map type has nullability issue for its nested key and values, because currently we create the map type and ignore its key/value nullability (both of them are from the type information with nullability true as default).

Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/schema/MapRelDataType.scala:48]: related issue: https://issues.apache.org/jira/browse/FLINK-20785

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Review by godfreyhe — APPROVED** [review]
